### DISCUSSION TITLE
Able to connect and validate my flight software using fprime-gds but a similar test using fprime-cli fails to connect

### DISCUSSION THREAD

**Discussion by ReggieMarr** [nasa/fprime#2525] [Q&A] [answered]
Hi all, I'm working on an existing project which uses fprime as the framework for our flight computer software and I'm struggling to understand the differences between running the code directly via running the binary directly and then articulating it via the fprime-cli vs running and articulating via the fprime-gds.

Specifically, when running the flight software I do so like this:

```
./FlightComputer/build-artifacts/Linux/bin/FlightComputer -a 127.0.0.1 -p 60000 -d 50000

```
Which yields the following output:
```
❯ ./FlightComputer/build-artifacts/Linux/bin/FlightComputer -a 127.0.0.1 -p 60000 -d 50000
Hit Ctrl-C to quit
// Event registration here
EVENT: (3334) (2:1707842809,645418) WARNING_HI: (prmDb) PrmFileReadError : Parameter file read failed in stage OPEN with record 0 and error 1
EVENT: (3328) (2:1707842809,645425) WARNING_LO: (prmDb) PrmIdNotFound : Parameter ID 0x4700 not found
EVENT: (3328) (2:1707842809,645428) WARNING_LO: (prmDb) PrmIdNotFound : Parameter ID 0x4701 not found
EVENT: (3328) (2:1707842809,645432) WARNING_LO: (prmDb) PrmIdNotFound : Parameter ID 0x2600 not found
EVENT: (3328) (2:1707842809,645434) WARNING_LO: (prmDb) PrmIdNotFound : Parameter ID 0x2601 not found
EVENT: (3328) (2:1707842809,645438) WARNING_LO: (prmDb) PrmIdNotFound : Parameter ID 0xf00 not found

[WARNING] Failed to open port with status -4 and errno 111

EVENT: (512) (2:1707842809,645716) DIAGNOSTIC: (rateGroup1Comp) RateGroupStarted : Rate group started.
[WARNING] High task priority of 118 being clamped to 99
EVENT: (768) (2:1707842809,645744) DIAGNOSTIC: (rateGroup2Comp) RateGroupStarted : Rate group started.
EVENT: (1024) (2:1707842809,645754) DIAGNOSTIC: (rateGroup3Comp) RateGroupStarted : Rate group started.
[WARNING] High task priority of 100 being clamped to 99
Starting downlink UDP connection with ip 127.0.0.1 and port 50000nSetup to send udp to 127.0.0.1:50000
```

If I attempt to connect using the fprime-cli like so I get an error:

```
flightcomputer-software/FlightComputer on  transmitter-if [!?⇡] via △ v3.22.1 
❯ fprime-cli channels -p 50000 -c dummyCounter 
zsh: correct 'dummyCounter' to 'DummyCounter' [nyae]? n
[INFO] Autodetected artifacts root '/home/reggiemarr/Projects/flightcomputer-software/FlightComputer/build-artifacts' from deployment settings.ini file.
Traceback (most recent call last):
  File "/home/reggiemarr/.local/lib/python3.10/site-packages/fprime_gds/common/transport.py", line 213, in connect
    self.sock.connect((host, port))
ConnectionRefusedError: [Errno 111] Connection refused

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/reggiemarr/.local/bin/fprime-cli", line 8, in <module>
    sys.exit(main())
  File "/home/reggiemarr/.local/lib/python3.10/site-packages/fprime_gds/executables/fprime_cli.py", line 412, in main
    function(**argument_dict)
  File "/home/reggiemarr/.local/lib/python3.10/site-packages/fprime_gds/executables/fprime_cli.py", line 233, in command_func
    channels.ChannelsCommand.handle_arguments(*args, **kwargs)
  File "/home/reggiemarr/.local/lib/python3.10/site-packages/fprime_gds/common/gds_cli/base_commands.py", line 235, in handle_arguments
    cls._execute_query(connection_info, search_info, kwargs["timeout"])
  File "/home/reggiemarr/.local/lib/python3.10/site-packages/fprime_gds/common/gds_cli/base_commands.py", line 188, in _execute_query
    pipeline, api = test_api_utils.initialize_test_api(
  File "/home/reggiemarr/.local/lib/python3.10/site-packages/fprime_gds/common/gds_cli/test_api_utils.py", line 40, in initialize_test_api
    pipeline.connect(server_ip, server_port)
  File "/home/reggiemarr/.local/lib/python3.10/site-packages/fprime_gds/common/pipeline/standard.py", line 155, in connect
    self.client_socket.connect(connection_uri, incoming_tag, outgoing_tag)
  File "/home/reggiemarr/.local/lib/python3.10/site-packages/fprime_gds/common/transport.py", line 221, in connect
    raise TransportationException(
fprime_gds.common.transport.TransportationException: Failed to connect to transportation layer at: 127.0.0.1:50000. [Errno 111] Connection refused
^CException ignored in: <module 'threading' from '/usr/lib/python3.10/threading.py'>
Traceback (most recent call last):
  File "/usr/lib/python3.10/threading.py", line 1567, in _shutdown
    lock.acquire()
KeyboardInterrupt: 

flightcomputer-software/FlightComputer on  transmitter-if [!?⇡] via △ v3.22.1 
❯ fprime-cli channels -p 60000 -c dummyCounter 
zsh: correct 'dummyCounter' to 'DummyCounter' [nyae]? n
[INFO] Autodetected artifacts root '/home/reggiemarr/Projects/flightcomputer-software/FlightComputer/build-artifacts' from deployment settings.ini file.
Traceback (most recent call last):
  File "/home/reggiemarr/.local/lib/python3.10/site-packages/fprime_gds/common/transport.py", line 213, in connect
    self.sock.connect((host, port))
ConnectionRefusedError: [Errno 111] Connection refused

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/reggiemarr/.local/bin/fprime-cli", line 8, in <module>
    sys.exit(main())
  File "/home/reggiemarr/.local/lib/python3.10/site-packages/fprime_gds/executables/fprime_cli.py", line 412, in main
    function(**argument_dict)
  File "/home/reggiemarr/.local/lib/python3.10/site-packages/fprime_gds/executables/fprime_cli.py", line 233, in command_func
    channels.ChannelsCommand.handle_arguments(*args, **kwargs)
  File "/home/reggiemarr/.local/lib/python3.10/site-packages/fprime_gds/common/gds_cli/base_commands.py", line 235, in handle_arguments
    cls._execute_query(connection_info, search_info, kwargs["timeout"])
  File "/home/reggiemarr/.local/lib/python3.10/site-packages/fprime_gds/common/gds_cli/base_commands.py", line 188, in _execute_query
    pipeline, api = test_api_utils.initialize_test_api(
  File "/home/reggiemarr/.local/lib/python3.10/site-packages/fprime_gds/common/gds_cli/test_api_utils.py", line 40, in initialize_test_api
    pipeline.connect(server_ip, server_port)
  File "/home/reggiemarr/.local/lib/python3.10/site-packages/fprime_gds/common/pipeline/standard.py", line 155, in connect
    self.client_socket.connect(connection_uri, incoming_tag, outgoing_tag)
  File "/home/reggiemarr/.local/lib/python3.10/site-packages/fprime_gds/common/transport.py", line 221, in connect
    raise TransportationException(
fprime_gds.common.transport.TransportationException: Failed to connect to transportation layer at: 127.0.0.1:60000. [Errno 111] Connection refused
```

But when I use fprime-gds things work just fine (and I can verify data is coming through on the gui as expected)

```
❯ fprime-gds                                   
[INFO] Autodetected artifacts root '/home/reggiemarr/Projects/flightcomputer-software/FlightComputer/build-artifacts' from deployment settings.ini file.
[INFO] Ensuring TCP Server is stable for at least 5 seconds
[INFO] Running Application: /usr/bin/python3
[INFO] Log File: /home/reggiemarr/Projects/flightcomputer-software/FlightComputer/logs/2024_02_13-11_55_55/ThreadedTCP.log
[INFO] Ensuring comm[ip] Application is stable for at least 1 seconds
[INFO] Running Application: /usr/bin/python3
[INFO] Ensuring FlightComputer Application is stable for at least 1 seconds
[INFO] Running Application: /home/reggiemarr/Projects/flightcomputer-software/FlightComputer/build-artifacts/Linux/bin/FlightComputer
[INFO] Log File: /home/reggiemarr/Projects/flightcomputer-software/FlightComputer/logs/2024_02_13-11_55_55/FlightComputer.log
[INFO] Ensuring HTML GUI is stable for at least 2 seconds
[INFO] Running Application: /usr/bin/python3
 * Serving Flask app "fprime_gds.flask.app"
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: off
[INFO] F prime is now running. CTRL-C to shutdown all components.

```

So whats going on here ? How is fprime gds running the application differently?

**Comment by thomas-bc** [comment, marked as answer]
`fprime-cli` needs to connect to the ground system layer, which is started by running `fprime-gds`. If you don't want to be using the web browser at all, you can start with `fprime-gds --gui none` to have no GUI.

So the process to use fprime-cli would be something like ...
```
fprime-gds &
fprime-cli channels
```
... or run in two terminals.

You can also find more information in the [User’s Guide to the F´ GDS CLI](https://nasa.github.io/fprime/UsersGuide/gds/gds-cli.html)

**Reply by ReggieMarr** [reply]
Oh I see, in the guide where it talks about "Running the cli by itself" I thought it was stand alone.
Basically I wanted to write some integration test scripts (that could be run by my ci/cd) so I suppose I would just run fprime-gds --gui none and then seperately use fprime-cli as you've suggested ?

Could you shed a bit more light on how fprime-gds runs the flight software ?  Is it possible to get it to run the software by passing specific command line args? or would it be more canonical to leverage send-command to do stuff like set the downlink/up link port at run time?

**Reply by thomas-bc** [reply]
1. Yes that's probably how I would do it, `fprime-gds --gui none` in the background. However, if what you're writing is integration tests, you may want to look at [GDS Integration Test API User's Guide](https://nasa.github.io/fprime/UsersGuide/dev/testAPI/user_guide.html). This is a Python API that lets you write integration tests rather easily, with timeouts/counts/awaits etc.... The guide is a little out of date unfortunately, it's now much eaiser to set up - see https://github.com/fprime-community/fprime-workshop-led-blinker/blob/main/Components/Led/test/int/led_integration_tests.py.

2. To run the flight software with specific command line args, I would just run the gds with the `-n` option which **doesn't** start the flight software. Then, in another terminal, start your flight software manually with your own CLI args, and just make sure the ports align.
e.g.
```
# Terminal 1
fprime-gds -n

# Terminal 2
./FlightComputer/build-artifacts/Linux/bin/FlightComputer -a 127.0.0.1 -p 50000 <other args>
```

**Reply by ReggieMarr** [reply]
Oh wow, I don't know how I didn't come across the integration tests before but thanks ! when you say its out of date how could I go about checking which version of fprime-gds the python api's documentation maps to? for reference I'm using fprime-util 3.1.1

**Reply by thomas-bc** [reply]
I'm not sure that's information we have unfortunately... but you're probably good with 3.1.1. 
The API shouldn't have changed too much, I think it's mostly the setup that was simplified using pytest fixtures. It should be able to automatically find a running fprime-gds so you don't have to configure it in your code. Or something like that.... @LeStarch can you confirm?

We have an issue to update this part of the docs (https://github.com/nasa/fprime/issues/1902). Feel free to make contributions by updating that as you walk through it if you'd like, this would be greatly appreciated!

**Reply by thomas-bc** [reply]
Ah nevermind, the pytest fixture was merged post 3.1.1 (https://github.com/fprime-community/fprime-gds/pull/109), so you may not have access to it and the current documentation is probably dating from about your version. 

You might be able to upgrade the GDS without breaking anything though, if you're comfortable with that and want to give it a spin. Up to you.

**Reply by ReggieMarr** [reply]
I think I'll likely stabilize my existing features set on v3.1.1 then but yeah I do intend on reviewing an upgrade so when the time comes I'll make a MR with an bits of documentation that appear missing

**Reply by ReggieMarr** [reply]
One more question, I noticed that in the documentation there's a reference to some run_ref_for_int_test.sh script which seems to no longer exist.
I'm putting together a docker-compose file plus some scripts that basically will startup the application however I want to make sure I'm not just re-doing something that is already handled.

**Reply by Joshua-Anderson** [reply]
Looks like the documentation might be out of date. Where did you see `run_ref_for_int_test.sh`? If you're looking for examples this is how integration tests are run in ci: https://github.com/nasa/fprime/blob/75b30ad8bb56bc65f55f3b05619b52b3acce8230/ci/tests/fputil.bash#L62-L129

**Reply by ReggieMarr** [reply]
There's a hyperlink to it under "Quick start guide" in the [api docs](https://github.com/nasa/fprime/blob/75b30ad8bb56bc65f55f3b05619b52b3acce8230/docs/UsersGuide/dev/testAPI/user_guide.md)

I can see that the approach that was taken was to spin up multiple processes in whatever env that instance is running in. That's what I did initially however I wanted to spin up multiple seperate docker containers so that I could leverage docker compose to control when things come up/health checks/docker logs ect.

My docker-compose wasn't working however and I thought I'd put together a shell script that brings up the different docker containers in a more manual fashion for now (just to prove to myself that the basic functionality is there)
```
#!/bin/bash
set -o nounset # Exit if the script tries to use undeclared variables

# Function to stop and remove the Docker container
stop_remove_container() {
    local CONTAINER_NAME=$1

    if [ "$(docker ps -aq -f name=${CONTAINER_NAME})" ]; then
        echo "Stopping ${CONTAINER_NAME} ..."
        docker stop ${CONTAINER_NAME}
        stop_string="Waiting for ${CONTAINER_NAME} to stop: "

        # Wait for the container to stop
        for i in {1..5}
        do
            if [ "$(docker ps -aq -f name=${CONTAINER_NAME})" ]; then
                echo "$i Waiting for ${CONTAINER_NAME} to stop..."
                sleep 0.2
            else
                break
            fi
        done

        if [ "$(docker ps -aq -f name=${CONTAINER_NAME})" ]; then
            echo "${CONTAINER_NAME} did not stop in time. Forcing removal ..."
            docker rm --force ${CONTAINER_NAME}
        else
            echo "Removing ${CONTAINER_NAME} ..."
            docker rm ${CONTAINER_NAME}
        fi
    else
        echo "Container ${CONTAINER_NAME} does not exist."
    fi
}

BUILD_CLEAN=0
SET_THREAD_CTRL=0

# gets flags
for arg in "$@"; do
  case $arg in
    --clean)
      BUILD_CLEAN=1
      ;;
    --host-thread-ctl)
      SET_THREAD_CTRL=1
      ;;
    *)
  esac
done

SCRIPT_DIR="$(cd "$(dirname "$0")"; pwd -P)"
FLIGHT_SW_IMAGE="git.reactiondynamics.space:5050/rdx-avionics-software/flightcomputer-software"
FLIGHT_SW_CONTAINER="flight-sw"
NETWORK_NAME=${NETWORK_NAME:-"flight-computer-net"}
COMMON_IP="192.168.96.1"

DOWNLINK_TARGET_IP="flight-sw-gds"
DOWNLINK_TARGET_PORT="50050"
UPLINK_TARGET_PORT="50000"
DEBUGPY_PORT="5678"

DFLAGS="--mount type=bind,source=${SCRIPT_DIR},target=/home/fprime/flightcomputer-software --network=${NETWORK_NAME} "
DFLAGS=$DFLAGS"-u $(id -u):$(id -g) "
BUILD_FLAGS=$DFLAGS"-w /home/fprime/flightcomputer-software/FlightComputer "
RUN_FLAGS=$DFLAGS"-w /home/fprime/flightcomputer-software/FlightComputer/build-artifacts/Linux/bin/ "
# RUN_FLAGS=$RUN_FLAGS"-p ${DOWNLINK_TARGET_PORT}:${DOWNLINK_TARGET_PORT} "
# The `--cap-add=SYS_NICE` flag allows processes to increase priorities and to set any CPU affinity.
# The `--ulimit rtprio=99` flag is setting the limit of real-time priority a user can set to 99, highest value for a typical kernel build.
RUN_FLAGS=$RUN_FLAGS"--cap-add=SYS_NICE --ulimit rtprio=99 "
BUILD_FLIGHT_SW="fprime-util build"
GEN_FLIGHT_SW="fprime-util purge --force && fprime-util generate && "$BUILD_FLIGHT_SW
RUN_FLIGHT_SW="./FlightComputer -a ${DOWNLINK_TARGET_IP} -u ${UPLINK_TARGET_PORT} -d ${DOWNLINK_TARGET_PORT}"
GDB_PORT=5555

case $1 in
  "teardown")
    if [ -n "$2" ]; then
        # $2 provided, so remove specific container
        stop_remove_container "${2}"
    else
        # No $2 provided, so remove all containers
        stop_remove_container "${FLIGHT_SW_CONTAINER}"
        stop_remove_container "${FLIGHT_SW_CONTAINER}-inspect"
        stop_remove_container "${FLIGHT_SW_CONTAINER}-gds"
    fi
    ;;
  "build")
    START_MODE=${START_MODE:-"-it --rm"}
    if [ "${BUILD_CLEAN}" -eq "1" ]; then
      echo 'Will clean and regnerate fpp derived files'
      CMD="${GEN_FLIGHT_SW}"
    else
      echo 'Performing build...'
      CMD="${BUILD_FLIGHT_SW}"
    fi
    docker run $START_MODE ${BUILD_FLAGS} --name ${FLIGHT_SW_CONTAINER} ${FLIGHT_SW_IMAGE} bash -c "${CMD}"
    cp -f $SCRIPT_DIR/FlightComputer/build-fprime-automatic-native/compile_commands.json $SCRIPT_DIR

    if [ "${SET_THREAD_CTRL}" -eq "1" ]; then
        echo 'Setting thread control for non-sudo host execution'
        sudo setcap 'cap_sys_nice+ep' ${SCRIPT_DIR}/FlightComputer/build-artifacts/Linux/bin/FlightComputer
    fi
    ;;
  "inspect")
    echo "Running FlightComputer interactively..."
    START_MODE=${START_MODE:-"-it --rm"}
    docker run $START_MODE ${BUILD_FLAGS} --name "${FLIGHT_SW_CONTAINER}-inspect" ${FLIGHT_SW_IMAGE} bash
    ;;
  "start-gds")
    echo "Running FlightComputer interactively..."
    BUILD_FLAGS=$BUILD_FLAGS"-p 5000:5000 "
    START_MODE=${START_MODE:-"-d"}
    GDS_CMD="fprime-gds -n --ip-address flight-sw-cli"
    docker run $START_MODE ${BUILD_FLAGS} --name "${FLIGHT_SW_CONTAINER}-gds" ${FLIGHT_SW_IMAGE} $GDS_CMD
    ;;
  "test-cli")
    echo "Running FlightComputer interactively..."
    START_MODE=${START_MODE:-"-it --rm"}
    CLI_CMD="fprime-cli channels -ip flight-sw-gds -c dummyCounter"
    docker run $START_MODE ${BUILD_FLAGS} --name "${FLIGHT_SW_CONTAINER}-cli" ${FLIGHT_SW_IMAGE} $CLI_CMD
    ;;
  "run-int-tests")
    echo "Running FlightComputer interactively..."
    START_MODE=${START_MODE:-"-it --rm"}
    INT_CMD="pytest -s ./test/int/flightcomputer_integration_test.py"
    docker run $START_MODE ${BUILD_FLAGS} --name "${FLIGHT_SW_CONTAINER}-gds-int" ${FLIGHT_SW_IMAGE} $INT_CMD
    ;;
  "script-run")
      shift # to get the next argument
      # Check if the script file exists
      script_to_run_in_docker=$1
      echo $script_to_run_in_docker
      START_MODE=${START_MODE:-"-it --rm"}
      docker exec -w /home/fprime/flightcomputer-software/FlightComputer ${FLIGHT_SW_CONTAINER} bash $script_to_run_in_docker
    ;;
  "debug")
    echo "Running FlightComputer with gdbserver..."
    RUN_FLAGS=$RUN_FLAGS"-p ${GDB_PORT}:${GDB_PORT} --cap-add=SYS_PTRACE --security-opt seccomp=unconfined "
    START_MODE=${START_MODE:-"-d"}
    echo $BUILD_FLAGS
    docker run $START_MODE ${RUN_FLAGS} --name ${FLIGHT_SW_CONTAINER} ${FLIGHT_SW_IMAGE} gdbserver :${GDB_PORT} ${RUN_FLIGHT_SW}
    ;;
  "run")
    echo "Running FlightComputer..."
    START_MODE=${START_MODE:-"-d"}
    echo "docker run $START_MODE ${RUN_FLAGS} --name ${FLIGHT_SW_CONTAINER} ${FLIGHT_SW_IMAGE} ${RUN_FLIGHT_SW}"
    docker run $START_MODE ${RUN_FLAGS} --name ${FLIGHT_SW_CONTAINER} ${FLIGHT_SW_IMAGE} ${RUN_FLIGHT_SW}
    ;;
  *)
    echo "Invalid operation. Please provide either 'build', 'run', or 'both'."
    exit 1
    ;;
esac

```
```
❯ ./run.sh run                                                                   
Running FlightComputer...
docker run -d --mount type=bind,source=/home/reggiemarr/Projects/flightcomputer-software,target=/home/fprime/flightcomputer-software --network=flight-computer-net -u 1000:1000 -w /home/fprime/flightcomputer-software/FlightComputer/build-artifacts/Linux/bin/ --cap-add=SYS_NICE --ulimit rtprio=99  --name flight-sw git.reactiondynamics.space:5050/rdx-avionics-software/flightcomputer-software ./FlightComputer -a flight-sw-gds -u 50000 -d 50050
1d905b850db7faf50572b081723dee6f6338d957630cf14856380ded4b1271d2

flightcomputer-software on  transmitter-if [$!?⇡] 
❯ docker logs -f flight-sw                                                       
Hit Ctrl-C to quit
EVENT: (1280) (2:1708055186,629377) DIAGNOSTIC: (cmdDisp) OpCodeRegistered : Opcode 0x2100 registered to port 0 slot 0
EVENT: (1280) (2:1708055186,629387) DIAGNOSTIC: (cmdDisp) OpCodeRegistered : Opcode 0x2101 registered to port 0 slot 1
EVENT: (1280) (2:1708055186,629390) DIAGNOSTIC: (cmdDisp) OpCodeRegistered : Opcode 0x2102 registered to port 0 slot 2
EVENT: (1280) (2:1708055186,629392) DIAGNOSTIC: (cmdDisp) OpCodeRegistered : Opcode 0x2200 registered to port 1 slot 3
EVENT: (1280) (2:1708055186,629394) DIAGNOSTIC: (cmdDisp) OpCodeRegistered : Opcode 0x2201 registered to port 1 slot 4
EVENT: (1280) (2:1708055186,629396) DIAGNOSTIC: (cmdDisp) OpCodeRegistered : Opcode 0x2202 registered to port 1 slot 5
EVENT: (1280) (2:1708055186,629398) DIAGNOSTIC: (cmdDisp) OpCodeRegistered : Opcode 0x2300 registered to port 2 slot 6
EVENT: (1280) (2:1708055186,629400) DIAGNOSTIC: (cmdDisp) OpCodeRegistered : Opcode 0x2301 registered to port 2 slot 7
EVENT: (1280) (2:1708055186,629402) DIAGNOSTIC: (cmdDisp) OpCodeRegistered : Opcode 0x2302 registered to port 2 slot 8
EVENT: (1280) (2:1708055186,629404) DIAGNOSTIC: (cmdDisp) OpCodeRegistered : Opcode 0x2400 registered to port 3 slot 9
EVENT: (1280) (2:1708055186,629407) DIAGNOSTIC: (cmdDisp) OpCodeRegistered : Opcode 0x2401 registered to port 3 slot 10
EVENT: (1280) (2:1708055186,629409) DIAGNOSTIC: (cmdDisp) OpCodeRegistered : Opcode 0x2402 registered to port 3 slot 11
EVENT: (1280) (2:1708055186,629411) DIAGNOSTIC: (cmdDisp) OpCodeRegistered : Opcode 0x2500 registered to port 4 slot 12
EVENT: (1280) (2:1708055186,629413) DIAGNOSTIC: (cmdDisp) OpCodeRegistered : Opcode 0x2501 registered to port 4 slot 13
EVENT: (1280) (2:1708055186,629415) DIAGNOSTIC: (cmdDisp) OpCodeRegistered : Opcode 0x2502 registered to port 4 slot 14
EVENT: (3334) (2:1708055186,629536) WARNING_HI: (prmDb) PrmFileReadError : Parameter file read failed in stage OPEN with record 0 and error 1
EVENT: (3328) (2:1708055186,629544) WARNING_LO: (prmDb) PrmIdNotFound : Parameter ID 0x4700 not found
EVENT: (3328) (2:1708055186,629547) WARNING_LO: (prmDb) PrmIdNotFound : Parameter ID 0x4701 not found
EVENT: (3328) (2:1708055186,629551) WARNING_LO: (prmDb) PrmIdNotFound : Parameter ID 0x2600 not found
EVENT: (3328) (2:1708055186,629554) WARNING_LO: (prmDb) PrmIdNotFound : Parameter ID 0x2601 not found
EVENT: (3328) (2:1708055186,629556) WARNING_LO: (prmDb) PrmIdNotFound : Parameter ID 0xf00 not found
[WARNING] High task priority of 140 being clamped to 99
[WARNING] High task priority of 101 being clamped to 99
[WARNING] High task priority of 100 being clamped to 99
[WARNING] High task priority of 100 being clamped to 99
[WARNING] High task priority of 100 being clamped to 99
[WARNING] Failed to open port with status -3 and errno 0
[WARNING] High task priority of 100 being clamped to 99
[ERROR] Failed to send framed data: 1
[ERROR] Failed to send framed data: 1
[ERROR] Failed to send framed data: 1
[ERROR] Failed to send framed data: 1
[ERROR] Failed to send framed data: 1
[ERROR] Failed to send framed data: 1
[WARNING] High task priority of 100 being clamped to 99
[WARNING] High task priority of 100 being clamped to 99
[WARNING] High task priority of 120 being clamped to 99
[WARNING] High task priority of 119 being clamped to 99
EVENT: (512) (2:1708055186,629903) DIAGNOSTIC: (rateGroup1Comp) RateGroupStarted : Rate group started.
[WARNING] High task priority of 118 being clamped to 99
EVENT: (768) (2:1708055186,629935) DIAGNOSTIC: (rateGroup2Comp) RateGroupStarted : Rate group started.
[WARNING] High task priority of 100 being clamped to 99
EVENT: (1024) (2:1708055186,629971) DIAGNOSTIC: (rateGroup3Comp) RateGroupStarted : Rate group started.
Starting downlink UDP connection with ip flight-sw-gds and port 50050n[ERROR] Failed to send framed data: 1
^C

flightcomputer-software on  transmitter-if [$!?⇡] 
❯ ./run.sh start-gds      
Running FlightComputer interactively...
d141df124d0a06abfd52a6d47aceb1fb29fbddb2d175d23f2239d2bfbfd54fbf

flightcomputer-software on  transmitter-if [$!?⇡] 
❯ docker logs -f flight-sw-gds
 * Serving Flask app "fprime_gds.flask.app"
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: off
^C

❯ docker network inspect flight-computer-net
[
    {
        "Name": "flight-computer-net",
        "Id": "f0f80f8562096495b3de00fde8f65251fb8cb2918f521672e55a993631c520c0",
        "Created": "2024-02-14T17:56:32.017269692-05:00",
        "Scope": "local",
        "Driver": "bridge",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": {},
            "Config": [
                {
                    "Subnet": "192.168.96.0/20",
                    "Gateway": "192.168.96.1"
                }
            ]
        },
        "Internal": false,
        "Attachable": false,
        "Ingress": false,
        "ConfigFrom": {
            "Network": ""
        },
        "ConfigOnly": false,
        "Containers": {
            "1d905b850db7faf50572b081723dee6f6338d957630cf14856380ded4b1271d2": {
                "Name": "flight-sw",
                "EndpointID": "95304b1f1e12a45913e7a73d564983d128bbf551e185088ed5fb82373602eac7",
                "MacAddress": "02:42:c0:a8:60:02",
                "IPv4Address": "192.168.96.2/20",
                "IPv6Address": ""
            },
            "8944764881b63c43fabb2673ddcdb33f140d2e7e6df7fe8308afee884847d55b": {
                "Name": "flight-sw-cli",
                "EndpointID": "7a133dd314366bdf8c9251398cadf336e0111cf99433eb4ce2326b497784ea5d",
                "MacAddress": "02:42:c0:a8:60:04",
                "IPv4Address": "192.168.96.4/20",
                "IPv6Address": ""
            },
            "d141df124d0a06abfd52a6d47aceb1fb29fbddb2d175d23f2239d2bfbfd54fbf": {
                "Name": "flight-sw-gds",
                "EndpointID": "100837729aee516ef35c85f6bbb72156645085615887d81a46a7b9a4cc804962",
                "MacAddress": "02:42:c0:a8:60:03",
                "IPv4Address": "192.168.96.3/20",
                "IPv6Address": ""
            }
        },
        "Options": {},
        "Labels": {}
    }
]
❯ ./run.sh test-cli
Running FlightComputer interactively...
[INFO] Autodetected artifacts root '/home/fprime/flightcomputer-software/FlightComputer/build-artifacts' from deployment settings.ini file.


```

Basically my test-cli seems to be hanging (this works when I just tell the flight computer to use the downlink ip local host and set the network to host) I think I'm misunderstanding something about how this all works together though.

My understanding is that the deployment binary sends data to the server running with ```fprime-gds -n``` invoked (especially since I set the downlink IP to be flight-sw-gds) and then I should be able to communicate with it via fprime-cli but for some reason it sucessfully connects to the gds but I can't seem to receive anything.

**Reply by Joshua-Anderson** [reply]
I'm not sure the FSW is connected to the GDS. The EVRs from the FSW seem to be indicating a failure to connect to the GDS. It looks like you may be passing a dns address instead of an ip when starting the FSW. I suspect the flight software isn't doing a DNS lookup, so it probably requires an actual IP address.

**Reply by ReggieMarr** [reply]
It seems you're correct, the hanging appears to occur when a connection is made to a valid IP (like localhost) but that ip doesn't actually have data coming in (which makes I suppose).

I have a docker-compose.yml that I'm currently using to perform a loopback test (I'm using this in place of a script very similar to [fputil](https://github.com/nasa/fprime/blob/75b30ad8bb56bc65f55f3b05619b52b3acce8230/ci/tests/fputil.bash#L62-L129)) would it make sense to make a PR with this ? Personally to me it seems that leveraging docker-compose to handle the gds, fsw, and integration test scripts as separate containers provides a bit more granularity in the control but thats just my preference.
